main.py

```diff
@@ -117,16 +104,44 @@ class MechaClient(Client):
     log.debug("starting bot for server...")
     try:
         log.debug("spawning new bot instance...")
-        client = MechaClient(IRC.presence)
-
-        log.info(f"connecting to {IRC.server}:{IRC.port}")
-        pool.connect(client, IRC.server, IRC.port, tls=IRC.tls)
+        if CONFIGURATION['authentication']['method'] == "PLAIN":
+            log.info("Authentication method set to PLAIN.")
+            # authenticate via sasl PLAIN mechanism (username & password)
+            client = MechaClient(CONFIGURATION['irc']['presence'],
+                                 sasl_username=CONFIGURATION['authentication']['plain']['username'],
+                                 sasl_password=CONFIGURATION['authentication']['plain']['password'],
+                                 sasl_identity=CONFIGURATION['authentication']['plain']['identity'])
+
+        elif CONFIGURATION['authentication']['method'] == "EXTERNAL":
+            log.info("Authentication method set to EXTERNAL")
+            # authenticate using provided client certificate
+            # key and cert may be stored as separate files, as long as mecha can read them.
+            cert = CONFIGURATION['authentication']['external']['tls_client_cert']
+            # key = CONFIGURATION['authentication']['external']['tls_client_key']
+
+            client = MechaClient(
+                CONFIGURATION['irc']['presence'],
+                sasl_mechanism='EXTERNAL',
+                tls_client_cert=f"certs/{cert}",
+                # tls_client_key=f"certs/{key}"
+            )
+        else:
+            # Pydle doesn't appear to support anything else
+            raise TypeError(f"unknown authentication mechanism "
+                            f"{CONFIGURATION['authentication']['method']}.\n"
+                            f"loading cannot continue.")
+
+        log.info(f"connecting to {CONFIGURATION['irc']['server']}:{CONFIGURATION['irc']['port']}")
+        pool.connect(client,
+                     CONFIGURATION['irc']['server'],
+                     CONFIGURATION['irc']['port'],
+                     tls=CONFIGURATION['irc']['tls'])
     except Exception as ex:
-        log.error(f"unable to connect to {IRC.server}:{IRC.port} "
+        log.error(f"unable to connect to {CONFIGURATION['irc']['server']}:"
+                  f"{CONFIGURATION['irc']['port']}"
                   f"due to an error.")
         log.error(ex)
-        from sys import exit
-        exit(42)
+        raise ex
```

Comment:
I feel somewhat that this except block is a bit useless. You're printing the error and then printing the exception only to re-raise the exception, which will see it printed out anyhoo.

Reply:
I think this can be useful for logging errors to either a database or file. Mecha will be run unattended so having it dump to a file will be valuable, but we still want to re-throw the error so mecha stops.

Reply:
Yeah, this method ensures the exception gets logged to disk. Since it is still a fatal error we want to re-raise it afterwards to ensure mecha crashes as fatal errors are not recoverable.